### ISSUE TITLE
KerberosClient.GetServiceTicket suddenly starts throwing System.InvalidOperationException

### ISSUE THREAD

**Issue by BrianBroe** [SteveSyfuhs/Kerberos.NET#216]
**Describe the bug**
After running fine for several day, suddenly the KerberClient starts throwing the following error when getting a ServiceTicket:

System.InvalidOperationException
> Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.

StackTrace:
```text
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Kerberos.NET.Crypto.CryptoService.ConvertType(EncryptionType type) in D:\a\1\s\Kerberos.NET\Crypto\CryptoService.cs:line 103
   at Kerberos.NET.Entities.KrbChecksum.Create(ReadOnlyMemory`1 data, KerberosKey key, KeyUsage ku, ChecksumType type) in D:\a\1\s\Kerberos.NET\Entities\Krb\KrbChecksum.cs:line 55
   at Kerberos.NET.Entities.KrbTgsReq.CreateTgsReq(RequestServiceTicket rst, KrbEncryptionKey tgtSessionKey, KrbKdcRep kdcRep, KrbEncryptionKey& sessionKey) in D:\a\1\s\Kerberos.NET\Entities\Krb\KrbTgsReq.cs:line 50
   at Kerberos.NET.Client.KerberosClient.RequestTgs(RequestServiceTicket rst, KerberosClientCacheEntry tgtEntry, CancellationToken cancellation) in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 910
   at Kerberos.NET.Client.KerberosClient.GetServiceTicket(RequestServiceTicket rst, CancellationToken cancellation) in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 521
   at Kerberos.NET.Client.KerberosClient.GetServiceTicket(String spn, ApOptions options, String s4u, KrbTicket s4uTicket, KrbTicket u2uServerTicket) in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 746
   at Api.HttpHandlers.KerberosTicketProvider`1.GetKerberosTicket(String serviceDomain) in /app/Api/HttpHandlers/KerberosTicketProvider.cs:line 51
   at Api.HttpHandlers.KerberosTicketProvider`1.GetKerberosTicket(String user, String password, String serviceDomain) in /app/Api/HttpHandlers/KerberosTicketProvider.cs:line 44
   at Api.HttpHandlers.KerberosHandler`1.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /app/Api/HttpHandlers/KerberosHandler.cs:line 29
```

The KerberosTicketProvider looks like this: 
```csharp
using System;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Net;
using System.Threading.Tasks;
using Kerberos.NET.Client;
using Kerberos.NET.Credentials;
using Microsoft.Extensions.Logging;
using Serilog;
using Serilog.Context;

namespace Api.HttpHandlers
{
    [ExcludeFromCodeCoverage]
    public class KerberosTicketProvider<T> : IKerberosTicketProvider<T>
    {
        private readonly KerberosClient _kerberosClient;
        private const string Domain = "DOMAIN.NET";

        public KerberosTicketProvider(ILoggerFactory factory)
        {
            _kerberosClient = new KerberosClient(logger:factory);
            var realmIpAddresses = Dns.GetHostAddresses(Domain);
            _kerberosClient.RenewTickets = true;
            _kerberosClient.PinKdc(Domain, realmIpAddresses.First().ToString());
        }

        public async Task<string> GetKerberosTicket(string user, string password, string serviceDomain)
        {
            var kerberosDetails = new {HttpClient = typeof(T).Name, User = user, ServiceDomain = serviceDomain};
            using (LogContext.PushProperty("KerberosDetails", kerberosDetails, true))
            {
                try
                {
                    return await GetKerberosTicket(serviceDomain);
                }
                catch (InvalidOperationException ex) //Thrown when user is not authenticated in KerberosClient.
                {
                    Log.Information(ex, "{User} was not authenticated on {ServiceDomain}", user, serviceDomain);
                }
                await Authenticate(user, password);
                return await GetKerberosTicket(serviceDomain);
            }
        }

        private async Task<string> GetKerberosTicket(string serviceDomain)
        {
            Log.Logger.Debug("Getting token for {ServiceDomain}", serviceDomain);
            var kerberosServiceTicket = await _kerberosClient.GetServiceTicket(serviceDomain);
            return Convert.ToBase64String(kerberosServiceTicket.EncodeGssApi().ToArray());
        }

        private async Task Authenticate(string user, string password)
        {
            Log.Logger.Information("Authenticating {User}", user);
            var kerberosCredentials = new KerberosPasswordCredential(user, password, Domain);
            await _kerberosClient.Authenticate(kerberosCredentials);
        }
    }
}
```
The KerberosTicketProvider is used in a KerberosHandler that looks like this: 
```csharp
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading;
using System.Threading.Tasks;

namespace Api.HttpHandlers
{
    public class KerberosHandler<T> : DelegatingHandler
    {
        private readonly string _username;
        private readonly string _password;
        private readonly string _serviceDomain;
        private readonly IKerberosTicketProvider<T> _kerberosTicketProvider;

        public KerberosHandler(string username, string password, string serviceDomain, IKerberosTicketProvider<T> kerberosTicketProvider)
        {
            _username = username;
            _password = password;
            _serviceDomain = serviceDomain;
            _kerberosTicketProvider = kerberosTicketProvider;
        }

        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            var ticket = await _kerberosTicketProvider.GetKerberosTicket(_username, _password, _serviceDomain);
            request.Headers.Authorization = new AuthenticationHeaderValue("Negotiate", ticket);

            return await base.SendAsync(request, cancellationToken);
        }
    }
}
```
The KerberosTicketProvider is registered as a Singleton, and thus the KerberosClient is re-used. Once the KerberosClient authentication expires, an exception is thrown and we re-authenticate.
The KerberosHandler is registered as a handler on the HttpClient like this:
```csharp
services.AddHttpClient<IMyClient, MyClient>()
   .AddHttpMessageHandler(x => new KerberosHandler<IMyClient>(mySettings.User, mySettings.Password, mySettings.KerberosServiceDomain, x.GetService<IKerberosTicketProvider<IMyClient>>()));
```

**To Reproduce**
NA

**Expected behavior**
No errors. :)

**Additional context**
We are using Kerberos.NET 4.5.45

The service runs in a docker image in OpenShift 3. It had been running for days with no problems, when it suddenly started throwing the error. Re-deploying the service fixed the issue.

**Comment by SteveSyfuhs** [comment]
That's an interesting behavior. There's a mapping in CryptoService that allows us to take an EType and find it's corresponding checksum type:

```csharp
private static readonly Dictionary<EncryptionType, ChecksumType> Cache = new Dictionary<EncryptionType, ChecksumType>();
```

Since it's purely a read operation I would expect that to always work even concurrently. 

Are you consistently seeing this error, or was it a one time thing? 

Once the error happened did you see this exception thrown every time a request was made, or just the once (guessing every time since the reboot made it go away)?

**Comment by SteveSyfuhs** [comment]
Nevermind. We do write it to add to the cache the first time through. I'll have a fix up shortly.

**Comment by SteveSyfuhs** [comment]
Fixed in https://github.com/dotnet/Kerberos.NET/pull/217.

**Comment by SteveSyfuhs** [comment]
This should be fixed in nuget 4.4.50: https://www.nuget.org/packages/Kerberos.NET/4.5.50

**Comment by BrianBroe** [comment]
Great! Thanks for the quick response. :)

Just to answer a few of your questions:
It has only gotten into this state once, but when it did, all of our attemtps to get a ticket from that KerberosClient failed, until we restarted.
I could not reproduce the error, so that't why I thought that restarting the service might help.

Do you know why we got into that state? As I said it had been running for a few days when it happened. 
Is it because the client is long living, and we re-authenticate? Should we dispose of the client and initialize a new when we do the authentication anyway?

**Comment by SteveSyfuhs** [comment]
Gotcha, thanks. That jives with the exception message. It's a global cache so reusing the client isn't the cause and switching to a per-request client wouldn't solve the problem either. Basically what's happening is a request comes in, causes the cache to miss so it goes through the matching logic and finds the value and stores it. The next 99 times out of a hundred (or whatever) this will be a read operation because it'll be the same etypes over and over again, but every one in a hundred will cause a cache miss because it's a different etype causing a single write. This write happens as a read is executing causing a (potential) corruption of the internal structures.

The exception itself is new. It was added sometime between .NET Core 3 and 5 and is thrown when two separate threads read and write at the same time. As such I think it only gets raised in the newest runtimes. According to the exception history this could lead to a corrupted internal list, which would "lead to an infinite loop" in the internal lookup, and to prevent it they now throw. That said, I've used this particular pattern consistently over the last few years in high volume apps so it's unclear why it's an issue now, but I digress. The fix was to switch to a `ConcurrentDictionary` which is designed for just this purpose.

In general the client is fine to stick around for long periods of time. There's an internal thread used to keep the cache cleaned up. The only hard and fast rule is one client per authenticated user.
